src/Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs

```diff
@@ -2683,12 +2707,19 @@ static bool IsWinRTType(ISymbol symbol)
             }
         }
 
-        public bool IsPublic(ISymbol type)
+        public bool IsPublic(ISymbol symbol)
         {
-            return type.DeclaredAccessibility == Accessibility.Public ||
-                type is IMethodSymbol method && !method.ExplicitInterfaceImplementations.IsDefaultOrEmpty ||
-                type is IPropertySymbol property && !property.ExplicitInterfaceImplementations.IsDefaultOrEmpty ||
-                type is IEventSymbol @event && !@event.ExplicitInterfaceImplementations.IsDefaultOrEmpty;
+            // Check that the type has either public accessibility, or is an explicit interface implementation
+            if (symbol.DeclaredAccessibility == Accessibility.Public ||
+                symbol is IMethodSymbol method && !method.ExplicitInterfaceImplementations.IsDefaultOrEmpty ||
+                symbol is IPropertySymbol property && !property.ExplicitInterfaceImplementations.IsDefaultOrEmpty ||
+                symbol is IEventSymbol @event && !@event.ExplicitInterfaceImplementations.IsDefaultOrEmpty)
+            {
+                // If we have a containing type, we also check that it's publicly accessible
+                return symbol.ContainingType is not { } containingType || containingType.IsPubliclyAccessible();

```

Comment:
I am trying to understand this, should this have been an `&&` rather than a `||`  or am I misunderstanding what is not {} does.

Reply:
So, `is not { }` matches if the expression is not an instance (it's like `is null`, but also declares a local). So:
- If `containingType` is `null`, return `true`
- Otherwise (ie. `containingType != null`), return `true` if it's publicly accessible

If we used `&&`, we'd be trying to access `containingType` when the first expression already matched, but if that's the case, then `containingType` is not defined (because the expression matches on `is not`). In fact, if you changed it to `&&`, the code would just not compile, as `containingType` would be uninitialized in that case 🙂